### PR TITLE
drivers/sensor: add support to LPS28DFW pressure sensor

### PR DESCRIPTION
**Pull request by jonas-rem**
Add support for the ST LPS28DFW pressure sensor. The sensor differs in the following points from the otherwise similar LPS22DF:

- Dual FS (260 to 1060 hPa and 260 to 4060 hPa)
- I2C and I3C, no SPI
- Waterproof (no SW consequences)

The approach of adding a new driver for this device has been discussed in #64766. This PR takes #62028 as a reference and implements the required changes (dual FS, no SPI).

~~@avisconti as there are not many changes, I did not change the copyright notice in the new driver files so it shows STMicroelecronics. Is this the correct approach?~~

### CODE REVIEW

**Comment by kartben** [comment]
> @avisconti as there are not many changes, I did not change the copyright notice in the new driver files so it shows STMicroelecronics. Is this the correct approach?

Multiple copyright lines are perfectly fine so feel free to add your own copyright entry too.

**Comment by avisconti** [comment]
> @avisconti as there are not many changes, I did not change the copyright notice in the new driver files so it shows STMicroelecronics. Is this the correct approach?

Exactly, I was to comment the same. Yo can add your own copyright while preserving the original copyright.

**Comment by avisconti** [comment]
> > @avisconti as there are not many changes, I did not change the copyright notice in the new driver files so it shows STMicroelecronics. Is this the correct approach?
> 
> Multiple copyright lines are perfectly fine so feel free to add your own copyright entry too.

Exactly, I was about to comment the same. Yo can add your own copyright while preserving the original copyright.

**Comment by jremmert-phytec-iot** [comment]
> > > @avisconti as there are not many changes, I did not change the copyright notice in the new driver files so it shows STMicroelecronics. Is this the correct approach?
> > 
> > 
> > Multiple copyright lines are perfectly fine so feel free to add your own copyright entry too.
> 
> Exactly, I was about to comment the same. Yo can add your own copyright while preserving the original copyright.

Thanks for your info @kartben and @avisconti ! Just updated the PR.

File: drivers/sensor/lps28dfw/lps28dfw.c [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/logging/log.h>
+
+#include "lps28dfw.h"
+
+#define LPS28DFW_SWRESET_WAIT_TIME	50
```
Comment (fabiobaltieri) [inline drivers/sensor/lps28dfw/lps28dfw.c:24]: `LPS28DFW_SWRESET_WAIT_TIME_US`

File: drivers/sensor/lps28dfw/lps28dfw.c [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+static inline void lps28dfw_press_convert(const struct device *dev,
+					 struct sensor_value *val,
+					 int32_t raw_val)
+{
+	int32_t press_tmp = raw_val >> 8; /* raw value is left aligned (24 msb) */
+	lps28dfw_md_t md;
+
+	md.fs = ((struct lps28dfw_config *)dev->config)->fs;
```
Comment (fabiobaltieri) [inline drivers/sensor/lps28dfw/lps28dfw.c:71]: how about adding an intermediate variable for `dev->config`? bit easier to read, you already do it in `lps28dfw_init_chip`

File: drivers/sensor/lps28dfw/lps28dfw.c [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+	.channel_get = lps28dfw_channel_get,
+#if CONFIG_LPS28DFW_TRIGGER
+	.trigger_set = lps28dfw_trigger_set,
+#endif
+};
+
+static int lps28dfw_init_chip(const struct device *dev)
+{
+	const struct lps28dfw_config * const cfg = dev->config;
+	__maybe_unused struct lps22df_data *data = dev->data;
```
Comment (fabiobaltieri) [inline drivers/sensor/lps28dfw/lps28dfw.c:173]: can you move this in the `if()` block where it's used? then you can do without `__maybe_unused`

**Comment by avisconti** [comment]
@jremmert-phytec-iot I told you before to replicate the LPS22DF driver, but to be honest I'm changing my mind and trying to think how to re-use the lps22df driver with lps28dfw extensions. I think that this can be done in the device tree itself adding in the dts bindings the compatible "st, lps28dfw".

dts/bindings/sensor/st,lps28dfw.yaml:

```
# Copyright (c) 2023, STMicroelectronics 
# (actually your copyright here)
# SPDX-License-Identifier: Apache-2.0

description: |
   STMicroelectronics LPS28DFW pressure and temperature sensor. This is an extension of st,lps22df driver binding.
   (other more intelligent description here...)

compatible: "st,lps28dfw"

(then attributes that extend the sensors if any)
```

In the lps22df driver then, you have to do something like:

```
#if DT_HAS_COMPAT_STATUS_OKAY(st_lps28dfw)
(extension code. Here, you can use the STdC APIs for lps28dfw)
#endif
```
I don't know if @fabiobaltieri agrees and can also help with some precious comment...

File: drivers/sensor/lps28dfw/lps28dfw.c (hunk 1 above)
Comment (avisconti) [reply drivers/sensor/lps28dfw/lps28dfw.c:24]: This was taken from lps22df, just merged. Probably we missed it during the review process.
I'm proposing to extend lps22df driver with lps28dfw stuff, which are not that much.

If agreed, we can go in this direction and @jremmert-phytec-iot may also follow up with these requests in lps22df driver.

**Comment by jremmert-phytec-iot** [comment]
> @jremmert-phytec-iot I told you before to replicate the LPS22DF driver, but to be honest I'm changing my mind and trying to think how to re-use the lps22df driver with lps28dfw extensions. I think that this can be done in the device tree itself adding in the dts bindings the compatible "st, lps28dfw".
> 
> dts/bindings/sensor/st,lps28dfw.yaml:
> 
> ```
> # Copyright (c) 2023, STMicroelectronics 
> # (actually your copyright here)
> # SPDX-License-Identifier: Apache-2.0
> 
> description: |
>    STMicroelectronics LPS28DFW pressure and temperature sensor. This is an extension of st,lps22df driver binding.
>    (other more intelligent description here...)
> 
> compatible: "st,lps28dfw"
> 
> (then attributes that extend the sensors if any)
> ```
> 
> In the lps22df driver then, you have to do something like:
> 
> ```
> #if DT_HAS_COMPAT_STATUS_OKAY(st_lps28dfw)
> (extension code. Here, you can use the STdC APIs for lps28dfw)
> #endif
> ```
> 
> I don't know if @fabiobaltieri agrees and can also help with some precious comment...

Yes, that should be a better solution. Will look into it later and test, either changing this PR or opening a second one so we have something to compare against.

File: drivers/sensor/lps28dfw/lps28dfw.c (hunk 1 above)
Comment (fabiobaltieri) [reply drivers/sensor/lps28dfw/lps28dfw.c:24]: yeah sorry, looks like I missed it

**Comment by fabiobaltieri** [comment]
> I don't know if @fabiobaltieri agrees and can also help with some precious comment...

Oh yeah totally, anything you could do to dedup the implementation, we don't really have a very coherent story for driver supporting multiple p/n in Zephyr unfortunately so you may have to compromise somewhere but it can be done, you may want to see at some examples and evaluate between messy but efficient macros or cleaner but less efficient [internal APIs](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/gpio/gpio_pcal64xxa.c#L84-L99).

**Comment by jremmert-phytec-iot** [comment]
@avisconti @fabiobaltieri would you mind having a quick look at https://github.com/jremmert-phytec-iot/zephyr/commit/3cc4696a78d95a01c2166f56381e51494e16b582? It's WIP and not yet ready for a PR, but I would like to know if I'm going in the right direction with it.

- Added the bindings for lps28dfw
- Extended Driver Instantiation macros to work with both `DT_DRV_COMPAT`
- I'm currently adding a chip_api and plan to keep a single driver_api
- naming is currently `lps2xdf`
- Builds and runs with two different device bindings activated

**Comment by fabiobaltieri** [comment]
@jremmert-phytec-iot I presume you need more than just `mode_set` in the `lps2xdf_chip_api` right? If it's just that the whole story could be simplified a fair bit I think, but the general approach loks right. Send a PR, commenting on commits is a bit messed up in GitHub.

**Comment by avisconti** [comment]
> @avisconti @fabiobaltieri would you mind having a quick look at [jremmert-phytec-iot@3cc4696](https://github.com/jremmert-phytec-iot/zephyr/commit/3cc4696a78d95a01c2166f56381e51494e16b582)? It's WIP and not yet ready for a PR, but I would like to know if I'm going in the right direction with it.
> 
> * Added the bindings for lps28dfw
> * Extended Driver Instantiation macros to work with both `DT_DRV_COMPAT`
> * I'm currently adding a chip_api and plan to keep a single driver_api
> * naming is currently `lps2xdf`
> * Builds and runs with two different device bindings activated

I'm adding some comments there...

**Comment by fabiobaltieri** [comment]
> I'm adding some comments there...

@jremmert-phytec-iot I'd really suggest opening a PR for it, even if you leave it as a draft. Not even sure you get notifications for commit comments.

**Comment by jremmert-phytec-iot** [comment]
@fabiobaltieri yes, the API is not yet complete. Thanks!
@avisconti as Fabio mentioned,  I'm also not able to see your comments. Sorry for making it harder than necessary to comment on it.

Sure, just opened a draft PR #65416. Don't feel obliged to do a full review right now. I'm happy with an approximate guidance at the current point.

**Pull request closed without merging**